plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,252 @@
+    their "unsaved" counterpart in the `~pulpcore.plugin.stages.DeclarativeArtifact` object.
+
+    Each `~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    `~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q: `~pulpcore.plugin.stages.DeclarativeContent`
+        out_q: `~pulpcore.plugin.stages.DeclarativeContent`
+
+    Returns:
+        The query_existing_artifacts stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content = []
+    shutdown = False
+    while True:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content:
+                content = await in_q.get()
+                declarative_content.append(content)
+                continue
+        else:
+            declarative_content.append(content)
+            continue
+
+        all_artifacts_q = Q(pk=None)
+        for content in declarative_content:
+            if content is None:
+                shutdown = True
+                continue
+
+            for declarative_artifact in content.d_artifacts:
+                one_artifact_q = Q()
+                for digest_name in declarative_artifact.artifact.DIGEST_FIELDS:
+                    digest_value = getattr(declarative_artifact.artifact, digest_name)
+                    if digest_value:
+                        key = {digest_name: digest_value}
+                        one_artifact_q &= Q(**key)
+                all_artifacts_q |= one_artifact_q
+
+        for artifact in Artifact.objects.filter(all_artifacts_q):
+            for content in declarative_content:
+                if content is None:
+                    continue
+                for declarative_artifact in content.d_artifacts:
+                    for digest_name in artifact.DIGEST_FIELDS:
+                        digest_value = getattr(declarative_artifact.artifact, digest_name)
+                        if digest_value and digest_value == getattr(artifact, digest_name):
+                            declarative_artifact.artifact = artifact
+        for content in declarative_content:
+            if content is None:
+                continue
+            await out_q.put(content)
+        declarative_content = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+class artifact_downloader:
+    """
+    An object containing a Stages API stage to download missing Artifacts, but don't call save()
+
+    The actual stage is the `stage` attribute which can be used as follows:
+
+    >>> artifact_downloader(max_concurrent_downloads=42).stage  # This is the real stage
+
+    in_q data type: A `~pulpcore.plugin.stages.DeclarativeContent` with potentially files missing
```

Comment:
The references to other docs are not resolving correctly in the built docs. 
![artifact_downloader_docs](https://user-images.githubusercontent.com/294339/43278243-83188cca-90d8-11e8-88f6-a5a3481bbd9d.png)

Reply:
This problem is all over the docs. I'm going through and fixing all of them with the next push.